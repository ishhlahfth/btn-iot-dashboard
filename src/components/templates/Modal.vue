<template>
  <transition name="fade" appear>
    <div
      v-if="modelValue"
      class="w-screen h-screen bg-black bg-opacity-50 fixed inset-0 z-50"
      @click="overlayClick"
    />
  </transition>
  <transition :name="screenWidth < 640 ? 'mobile-slide-up' : 'slide-up'" appear>
    <div
      v-if="modelValue"
      class="fixed modal-style p-4 sm:p-6 rounded-t-2xl sm:rounded-lg shadow-custom z-50 transition-colors duration-500 transform"
      :class="violated ? 'bg-flame-soft' : 'bg-snow'"
    >
      <slot></slot>
    </div>
  </transition>
</template>

<script>
export default {
  name: 'HelpModal',
  props: {
    modelValue: {
      type: Boolean,
      default: false,
    },
    permanent: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      violated: false,
    };
  },
  computed: {
    screenWidth() {
      return this.$store.state.screenWidth;
    },
  },
  methods: {
    overlayClick() {
      if (!this.permanent) {
        this.$emit('update:modelValue', false);
      } else {
        this.violated = true;
        setTimeout(() => {
          this.violated = false;
        }, 500);
      }
    },
  },
};
</script>

<style lang="scss">
.shadow-custom {
  box-shadow: 0px 1px 2px rgba(0, 0, 0, 0.15);
}

.slide-up-enter-active,
.slide-up-leave-active,
.mobile-slide-up-enter-active,
.mobile-slide-up-leave-active {
  transition: all 0.25s ease-in-out;
}

.slide-up-enter-from,
.slide-up-leave-to {
  top: 100vh !important;
  transform: translateY(100vh) !important;
  transform: translateX(-50%) !important;
}

.mobile-slide-up-enter-from,
.mobile-slide-up-leave-to {
  transform: translateY(100vh) !important;
}

.fade-enter-active,
.fade-leave-active {
  transition: all 0.5s ease-in-out;
}
.fade-enter-from,
.fade-leave-to {
  opacity: 0;
}
.modal-style {
  @apply bottom-0;
  @apply inset-x-0;
  @screen sm {
    @apply bottom-auto;
    @apply right-auto;
    @apply top-1/2;
    @apply left-1/2;
    transform: translate(-50%, -50%);
  }
  .inner-modal-fixed {
    height: 85vh;
  }
  .inner-modal-auto {
    height: auto;
  }
  .modal-xl {
    @screen sm {
      width: 85vw;
      height: 87.5vh;
    }
  }
  .modal-lg {
    @screen sm {
      width: 57.5vw;
      height: 75vh;
    }
  }
  .modal-md {
    @screen sm {
      min-width: 440px;
      max-width: 27.5vw;
      height: auto;
      max-height: 75.5vh;
    }
  }
  .modal-sm {
    @screen sm {
      min-width: 320px;
      max-width: 27.5vw;
      height: auto;
      max-height: 75.5vh;
    }
  }
}
</style>
