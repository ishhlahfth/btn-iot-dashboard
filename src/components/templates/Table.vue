<template>
  <div class="overflow-x-auto overflow-y-hidden rounded-lg border border-grey-4 divide-y-2">
    <div class="overflow-x-auto">
      <table class="min-w-full">
        <thead class="border-b border-grey-4 whitespace-nowrap min-w-full">
          <tr class="bg-white uppercase">
            <th
              v-for="(column, i) in columns"
              :key="i"
              class="py-3 px-6 text-small font-medium text-grey-2"
            >
              <p
                :class="[
                  { 'text-center': column.align === 'center' },
                  { 'text-right': column.align === 'right' },
                ]"
              >
                {{ column.label }}
              </p>
            </th>
          </tr>
        </thead>
        <tbody class="divide-y divide-grey-4 whitespace-nowrap">
          <tr v-for="(row, i) in rows" :key="i" class="bg-white">
            <td v-for="(data, i) in row" :key="i" class="py-3 px-6 text-small">
              <div
                class="w-full grid"
                :class="[
                  { 'place-items-center': columnAlignment(i) === 'center' },
                  { 'place-items-end': columnAlignment(i) === 'right' },
                ]"
              >
                <slot :column="i" :row="row" :data="data">
                  <!-- default, when there is no slot -->
                  <p>{{ data }}</p>
                </slot>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <table-footer />
  </div>
</template>

<script>
import TableFooter from '../molecules/TableFooter.vue';

export default {
  name: 'HelpTable',
  components: {
    TableFooter,
  },
  props: {
    columns: {
      type: Array,
      required: true,
      default() {
        return [];
      },
    },
    rows: {
      type: Array,
      required: true,
      default() {
        return [];
      },
    },
  },
  methods: {
    columnAlignment(columnName) {
      let alignment = '';
      console.log(columnName);
      const filtered = this.columns.filter((el) => el.field === columnName)[0];
      console.log(filtered, 'ðŸš€');
      if (filtered.align) {
        alignment = filtered.align;
      } else {
        alignment = 'left';
      }
      return alignment;
    },
  },
};
</script>

<style></style>
