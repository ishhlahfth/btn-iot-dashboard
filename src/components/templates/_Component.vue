<template>
  <section :id="id" class="grid gap-8 auto-rows-max">
    <p class="text-heading2 font-semibold text-midnight">{{ title }}</p>

    <slot name="description">-</slot>

    <div class="divide-y divide-grey-4 grid gap-2">
      <p class="font-semibold">Design</p>
      <!-- <div class="flex justify-between">
        <p class="font-semibold">Design</p>
        <help-button icon-only icon="code" bg-color="transparent" color="grey-1" />
      </div> -->
      <p></p>
    </div>

    <slot name="design">-</slot>

    <div class="divide-y divide-grey-4 grid gap-2">
      <p class="font-semibold">Props</p>
      <p></p>
    </div>
    <help-table
      :footer="false"
      :columns="[
        { field: 'prop', label: 'Prop' },
        { field: 'description', label: 'Description' },
        { field: 'type', label: 'Type' },
        { field: 'default', label: 'Default' },
        { field: 'examples', label: 'Examples' },
      ]"
      :rows="props"
    >
      <template v-slot:body="{ column, row }">
        <div v-if="column === 'examples'">
          <code
            v-for="(example, index) in row.examples"
            :key="index"
            class="highlight mr-1"
          >
            {{ example }}
          </code>
        </div>
      </template>
    </help-table>

    <template v-if="events?.length">
      <div class="divide-y divide-grey-4 grid gap-2">
        <p class="font-semibold">Events</p>
        <p></p>
      </div>
      <help-table
        :footer="false"
        :columns="[
          { field: 'event', label: 'Events' },
          { field: 'description', label: 'Description' },
          { field: 'parameters', label: 'Parameters' },
        ]"
        :rows="events"
      >
        <template v-slot:body="{ column, row }">
          <ol class="list-disc list-outside" v-if="column === 'parameters'">
            <li v-for="(param, index) in row.parameters" :key="index">
              <span class="highlight mr-1">{{ param.name }}</span>:
              <span>{{ param.param }}</span>
            </li>
          </ol>
        </template>
      </help-table>
    </template>

    <template v-if="slots?.length">
      <div class="divide-y divide-grey-4 grid gap-2">
        <p class="font-semibold">Slots</p>
        <p></p>
      </div>
      <help-table
        :footer="false"
        :columns="[
          { field: 'slot', label: 'slot name' },
          { field: 'description', label: 'description' },
          { field: 'props', label: 'slot props' },
        ]"
        :rows="slots"
      >
        <template v-slot:body="{ column, row }">
          <ol class="list-disc list-outside" v-if="column === 'props'">
            <li v-for="(prop, index) in row.props" :key="index">
              <span class="highlight mr-1">{{ prop.name }}</span>:
              <span>{{ prop.definition }}</span>
            </li>
          </ol>
        </template>
      </help-table>
    </template>

    <div v-if="hasUpdateGuideSlot" class="divide-y divide-grey-4 grid gap-2">
      <p class="font-semibold">Update Guide</p>
      <p></p>
    </div>
    <slot name="update-guide"></slot>
  </section>
</template>

<script>
import HelpTable from '@/components/templates/Table.vue';

export default {
  name: 'Component',
  props: ['id', 'title', 'props', 'events', 'slots'],
  components: {
    HelpTable,
  },
  computed: {
    hasUpdateGuideSlot() {
      return !!this.$slots['update-guide'];
    },
  },
};
</script>

<style></style>
