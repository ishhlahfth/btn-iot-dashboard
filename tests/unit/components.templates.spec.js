import _Component from '@/components/templates/_Component.vue';
import Modal from '@/components/templates/Modal.vue';
import Table from '@/components/templates/Table.vue';

describe('_Component.vue', () => {
  it('should render correctly', () => {
    expect(_Component.$el).toMatchSnapshot();
  });

  it('should render name Component', () => {
    expect(_Component.name).toBe('Component');
  });
});

describe('Modal.vue', () => {
  it('should render correctly', () => {
    expect(Modal.$el).toMatchSnapshot();
  });

  it('should render name Modal', () => {
    expect(Modal.name).toBe('HelpModal');
  });
});

describe('Table.vue', () => {
  it('should render correctly', () => {
    expect(Table.$el).toMatchSnapshot();
  });

  it('should render name Table', () => {
    expect(Table.name).toBe('HelpTable');
  });
});
