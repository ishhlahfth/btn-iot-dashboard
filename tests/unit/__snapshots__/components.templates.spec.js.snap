// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_Component.vue should render correctly 1`] = `undefined`;

exports[`Modal.vue should render correctly 1`] = `undefined`;

exports[`Table.vue should render correctly 1`] = `undefined`;
